ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java

```diff
@@ -0,0 +1,164 @@
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final EventBus eventBus;
+  private final ForkChoiceAttestationProcessor attestationProcessor;
+
+  private final PendingPool<ValidateableAttestation> pendingAttestations;
+  private final FutureItems<ValidateableAttestation> futureAttestations;
+  private final AggregatingAttestationPool aggregatingAttestationPool;
+
+  private final Subscribers<ProcessedAttestationConsumer> processedAttestationSubscriber =
+      Subscribers.create(true);
+  private final Subscribers<ProcessedAggregateConsumer> processedAggregateSubscriber =
+      Subscribers.create(true);
+
+  AttestationManager(
+      final EventBus eventBus,
+      final ForkChoiceAttestationProcessor attestationProcessor,
+      final PendingPool<ValidateableAttestation> pendingAttestations,
+      final FutureItems<ValidateableAttestation> futureAttestations,
+      final AggregatingAttestationPool aggregatingAttestationPool) {
+    this.eventBus = eventBus;
+    this.attestationProcessor = attestationProcessor;
+    this.pendingAttestations = pendingAttestations;
+    this.futureAttestations = futureAttestations;
+    this.aggregatingAttestationPool = aggregatingAttestationPool;
+  }
+
+  public static AttestationManager create(
+      final EventBus eventBus,
+      final PendingPool<ValidateableAttestation> pendingAttestations,
+      final FutureItems<ValidateableAttestation> futureAttestations,
+      final ForkChoiceAttestationProcessor forkChoiceAttestationProcessor,
+      final AggregatingAttestationPool aggregatingAttestationPool) {
+    return new AttestationManager(
+        eventBus,
+        forkChoiceAttestationProcessor,
+        pendingAttestations,
+        futureAttestations,
+        aggregatingAttestationPool);
+  }
+
+  public void subscribeToProcessedAttestations(
+      ProcessedAttestationConsumer processedAttestationConsumer) {
+    processedAttestationSubscriber.subscribe(processedAttestationConsumer);
+  }
+
+  public void subscribeToProcessedAggregates(
+      ProcessedAggregateConsumer processedAggregateConsumer) {
+    processedAggregateSubscriber.subscribe(processedAggregateConsumer);
+  }
+
+  @Override
+  public void onSlot(final UnsignedLong slot) {
+    List<ValidateableAttestation> attestations = futureAttestations.prune(slot);
+    attestations.stream()
+        .map(ValidateableAttestation::getIndexedAttestation)
+        .forEach(attestationProcessor::applyIndexedAttestationToForkChoice);
+
+    attestations.forEach(this::gossipAttestationIfNeeded);
+  }
+
+  private void gossipAttestationIfNeeded(ValidateableAttestation attestation) {
+    if (attestation.isGossiped()) {
+      return;
+    }
+
+    if (attestation.isAggregate()) {
+      processedAggregateSubscriber.forEach(s -> s.accept(attestation));
+    } else {
+      processedAttestationSubscriber.forEach(s -> s.accept(attestation));
+    }
```

Comment:
Not sure it makes sense to only notify listeners if the attestation hasn't been gossiped.  We don't know that the listener is gossiping it and we aren't marking it as gossiped.  Suspect here we should always notify subscribers, and then the subscriber that does the gossiping can check if it has already gossiped the attestation and ignore the notification if required.

This method would then just be `notifySubscribers`

Slightly tempted not to differentiate between single and aggregate attestations here either and again let consumers decide if they need to differentiate between the two.

Reply:
I was really on the line with both of your points here, and just did the way I did to make a decision and get feedback. Both points make sense to me.

Reply:
The second point especially makes sense since `AttestationManager` normally does not care about aggregate or not, and `ValidataeableAttestation` itself handles the difference. So it doesn't make much sense for `AttestationManager` to separate the two.